protocols/virtualfund/virtualfund_single_hop_test.go

```diff
@@ -0,0 +1,304 @@
+	// And then crank it to see which "pause point" (WaitingFor) we end up at.
+	// NOTE: Because crank returns a protocools.Objective interface, after each crank we
+	// need to remember to convert the result back to a virtualfund.Objective struct
+
+	// Initial Crank
+	oObj, got, waitingFor, err := o.Crank(&my.privateKey)
+	o = oObj.(*Objective)
+
+	expectedSignedState := state.NewSignedState(o.V.PreFundState())
+	mySig, _ := o.V.PreFundState().Sign(my.privateKey)
+	_ = expectedSignedState.AddSignature(mySig)
+
+	ok(t, err)
+	equals(t, waitingFor, WaitingForCompletePrefund)
+	assertStateSentTo(t, got, expectedSignedState, bob)
+	assertStateSentTo(t, got, expectedSignedState, p1)
+
+	// Manually progress the extended state by collecting prefund signatures
+	collectPeerSignaturesOnSetupState(o.V, my.role, true)
+
+	// Cranking should move us to the next waiting point, update the ledger channel, and alter the extended state to reflect that
+	// TODO: Check that ledger channel is updated as expected
+	oObj, got, waitingFor, err = o.Crank(&my.privateKey)
+	o = oObj.(*Objective)
+
+	p := consensus_channel.NewAddProposal(o.ToMyRight.Channel.Id, 2, o.ToMyRight.getExpectedGuarantee(), big.NewInt(6))
+	sp := consensus_channel.SignedProposal{Proposal: p}
+	assertProposalSent(t, got, sp, p1)
+	ok(t, err)
+	equals(t, waitingFor, WaitingForCompleteFunding)
+
+	// Check idempotency
+	emptySideEffects := protocols.SideEffects{}
+	oObj, got, waitingFor, err = o.Crank(&my.privateKey)
+	o = oObj.(*Objective)
+	ok(t, err)
+	equals(t, got, emptySideEffects)
+	equals(t, waitingFor, WaitingForCompleteFunding)
+}
+
+// Copied from https://github.com/benbjohnson/testing
+
+// assert fails the test if the condition is false.
+func assert(tb testing.TB, condition bool, msg string, v ...interface{}) {
+	if !condition {
+		_, file, line, _ := runtime.Caller(1)
+		fmt.Printf("\033[31m%s:%d: "+msg+"\033[39m\n\n", append([]interface{}{filepath.Base(file), line}, v...)...)
+		tb.FailNow()
+	}
+}
+
+// ok fails the test if an err is not nil.
+func ok(tb testing.TB, err error) {
+	if err != nil {
+		_, file, line, _ := runtime.Caller(1)
+		fmt.Printf("\033[31m%s:%d: unexpected error: %s\033[39m\n\n", filepath.Base(file), line, err.Error())
+		tb.FailNow()
+	}
+}
+
+// equals fails the test if exp is not equal to act.
+func equals(tb testing.TB, exp, act interface{}) {
+	if !reflect.DeepEqual(exp, act) {
+		_, file, line, _ := runtime.Caller(1)
+		fmt.Printf("\033[31m%s:%d:\n\n\texp: %#v\n\n\tgot: %#v\033[39m\n\n", filepath.Base(file), line, exp, act)
+		tb.FailNow()
+	}
+}
+
+// The following assertions are inspired by the ok, assert and equals above
+
+// assertSideEffectsContainsMessageWith fails the test instantly if the supplied side effects does not contain a message for the supplied actor with the supplied expected signed state.
+func assertProposalSent(t *testing.T, ses protocols.SideEffects, sp consensus_channel.SignedProposal, to actor) {
```

Comment:
It would be nice if formatting could be handled separately so `assertProposalSent` would be simpler to read.

Reply:
IMO, it's not very important to understand the formatting in `assertProposalSent`. It's only important that it works well :P

9e626058a6aac0e3ef2dbdae0243a115f4af0510 should make it simpler to read?